Add decision ack after commit/abort

vaporos opened the pull request:
Previously, the coordinator immediately went to the next epoch after
a decision was reached. This created a potential race condition in the
order of messages received and processed by a participant; if the next
epoch's RequestForVote message was received prior to the Commit/Abort
decision of the current epoch, then the RequestForVote message would be
dropped (because it wasn't for the current epoch). The result of the
dropped message was an eventual timeout and abort of that epoch.

This commit adds a new message sent from the participants to the
coordinator after a decision has been applied: DecisionAck. The
coordinator waits for all DecisionAck messages from all participants,
similar to collecting votes. Once all DecisionAcks are received, then
the coordinator starts the next epoch. If the coordinator times out, it
also starts the next epoch, since it is possible a decision ack will
never be sent by the processes that failed to respond.

vaporos committed:
Add decision ack after commit/abort

Previously, the coordinator immediately went to the next epoch after
a decision was reached. This created a potential race condition in the
order of messages received and processed by a participant; if the next
epoch's RequestForVote message was received prior to the Commit/Abort
decision of the current epoch, then the RequestForVote message would be
dropped (because it wasn't for the current epoch). The result of the
dropped message was an eventual timeout and abort of that epoch.

This commit adds a new message sent from the participants to the
coordinator after a decision has been applied: DecisionAck. The
coordinator waits for all DecisionAck messages from all participants,
similar to collecting votes. Once all DecisionAcks are received, then
the coordinator starts the next epoch. If the coordinator times out, it
also starts the next epoch, since it is possible a decision ack will
never be sent by the processes that failed to respond.

Signed-off-by: Shawn T. Amundson <amundson@bitwise.io>

agunde406 commented on libaugrim/src/two_phase_commit/coordinator_algorithm.rs:
```diff
@@ -29,6 +29,13 @@ use super::CoordinatorState;
 use super::TwoPhaseCommitContext;
 use super::TwoPhaseCommitMessage;
 
+// The timeout for receiving all decision acks before continuing to the next epoch.
```
typo in commit message: coordinaotr -> coordinator

vaporos replied:
fixed

vaporos committed:
Remove catch-all branches from state match statements

In TryFrom, previously we had a catch-all branch "_ =>" which handled
all states not covered. This is problematic when adjusting states since
the compiler won't force new states to be covered explicitly, leaving
room for coding errors. Thus, we avoid the catch-all here.

Signed-off-by: Shawn T. Amundson <amundson@bitwise.io>

vaporos commented:
Fixed match statement, added commit to remove related catch-all branches.
